src/metabase/query_processor/middleware/fetch_source_query.clj

```diff
@@ -94,42 +96,81 @@
+                             {:card-id card-id})))
+         {{mbql-query                   :query
+           database-id                  :database
+           {template-tags :template-tags
+            :as           native-query} :native} :dataset_query
+          result-metadata                        :result_metadata
+          dataset?                               :dataset}
+         card
+
+         [persisted? source-query]
+         (or (when (and persisted-info/*allow-persisted-substitution*
+                        (:active card)
+                        (:definition card)
+                        (:query_hash card)
+                        (= (:query_hash card) (persisted-info/query-hash (:dataset_query card)))
+                        (= (:definition card) (persisted-info/metadata->definition (:result_metadata card)
+                                                                                   (:table_name card)))
+                        (= (:state card) "persisted"))
+               (when log?
+                 (log/info (trs "Substituting cached query for card {0} from {1}.{2}"
+                                card-id
+                                (ddl.i/schema-name {:id database-id} (public-settings/site-uuid))
+                                (:table_name card))))
+               [true
+                {:native (format "select %s from %s.%s"
+                                 (str/join ", " (map :field-name (get-in card [:definition :field-definitions])))
+                                 (ddl.i/schema-name {:id database-id} (public-settings/site-uuid))
+                                 (:table_name card))}])
+             (when mbql-query
+               [false mbql-query])
+             (when native-query
+               ;; rename `:query` to `:native` because source queries have a slightly different shape
+               (let [native-query (set/rename-keys native-query {:query :native})]
+                 [false
+                  (cond-> native-query
+                    ;; trim trailing comments from SQL, but not other types of native queries
+                    (string? (:native native-query)) (update :native (partial trim-sql-query card-id))
+                    (empty? template-tags)           (dissoc :template-tags))]))
+             (throw (ex-info (tru "Missing source query in Card {0}" card-id)
+                             {:card card})))]
+     ;; log the query at this point, it's useful for some purposes
+     (log/debug (trs "Fetched source query from Card {0}:" card-id)
+                "\n"
+                (u/pprint-to-str 'yellow source-query))
+     (cond-> {:source-query    source-query
+              :database        database-id
+              :source-metadata (cond-> (seq (map mbql.normalize/normalize-source-metadata result-metadata))
+                                 persisted? sub-cached-field-refs)}
+       dataset? (assoc :source-query/dataset? dataset?)))))
```

Comment:
This function is 66 lines now might be time to try to break it out into smaller functions.